### PR TITLE
Dreambooth method for finetuning stable diffusions

### PR DESCRIPTION
**Pull request by XDUWQ**
Dreambooth is an effective way to fine-tune stable diffusion, the principle can refer to: https://arxiv.org/pdf/2208.12242.pdf

Only need to use 5-10 images to customize your stable diffusion. The dataset used in the example is buptwq/lora-stable-diffusion-finetune-dog.
image

After fine-tuning the dreambooth method, the generated dog's rendering is as follows.
image

### CODE REVIEW

File: modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+@TRAINERS.register_module(module_name=Trainers.dreambooth_diffusion)
+class DreamboothDiffusionTrainer(EpochBasedTrainer):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.with_prior_preservation = kwargs.pop('with_prior_preservation',
+                                                  False)
+        self.instance_prompt = kwargs.pop('instance_prompt',
+                                          'a photo of sks dog')
+        self.class_prompt = kwargs.pop('instance_prompt', 'a photo of dog')
```
Comment (wenmengzhou) [inline modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py:157]: is it right that class_prompt use instance_prompt value?

File: modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    A dataset to prepare  class images with the prompts for fine-tuning the model.
+    It pre-processes the images and the tokenizes prompts.
+    """
+
+    def __init__(
+        self,
+        tokenizer,
+        class_data_root=None,
+        class_prompt=None,
+        class_num=None,
```
Comment (wenmengzhou) [inline modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py:71]: class_num -> class_num_images

File: modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+        example['index'] = index
+        return example
+
+
+@TRAINERS.register_module(module_name=Trainers.dreambooth_diffusion)
+class DreamboothDiffusionTrainer(EpochBasedTrainer):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.with_prior_preservation = kwargs.pop('with_prior_preservation',
```
Comment (wenmengzhou) [inline modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py:153]: why with_prior_preservation is False by default

File: modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+        example = {}
+        example['prompt'] = self.prompt
+        example['index'] = index
+        return example
+
+
+@TRAINERS.register_module(module_name=Trainers.dreambooth_diffusion)
+class DreamboothDiffusionTrainer(EpochBasedTrainer):
+
+    def __init__(self, *args, **kwargs):
```
Comment (wenmengzhou) [inline modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py:151]: add docstr for each parameter

File: modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+                pipeline.set_progress_bar_config(disable=True)
+
+                num_new_images = self.num_class_images - cur_class_images
+                sample_dataset = PromptDataset(self.instance_prompt,
+                                               num_new_images)
+                sample_dataloader = torch.utils.data.DataLoader(sample_dataset)
+
+                pipeline.to(self.device)
+
+                for example in tqdm(
```
Comment (wenmengzhou) [inline modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py:209]: current implementation only support single gpu,  assertion should be used to check this condition,   and add multi-gpu support in the future

File: tests/pipelines/test_diffusers_stable_diffusion.py [hunk 6]
```diff
@@ -21,7 +21,7 @@ def setUp(self) -> None:
     def test_run(self):
         diffusers_pipeline = pipeline(task=self.task, model=self.model_id)
         output = diffusers_pipeline({
-            'prompt': self.test_input,
+            'text': self.test_input,
```
Comment (wenmengzhou) [inline tests/pipelines/test_diffusers_stable_diffusion.py:24]: prompt参数需要做一下兼容，保证老的代码调用不会出错

File: tests/trainers/test_dreambooth_diffusion_trainer.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            cfg_modify_fn=cfg_modify_fn)
+
+        trainer = build_trainer(
+            name=Trainers.dreambooth_diffusion, default_args=kwargs)
+        trainer.train()
+        result = trainer.evaluate()
+        print(f'Dreambooth-diffusion train output: {result}.')
+
+        results_files = os.listdir(self.tmp_dir)
+        self.assertIn(f'{trainer.timestamp}.log.json', results_files)
```
Comment (wenmengzhou) [inline tests/trainers/test_dreambooth_diffusion_trainer.py:68]: add test case for pipeline inference after finetuning

File: modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py (hunk 3 above)
Comment (XDUWQ) [reply modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py:167]: "With_prior_preservation true will preserve more diversity, while with_prior_preservation false will preserve more featureness."

File: modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py (hunk 5 above)
Comment (XDUWQ) [reply modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py:225]: Ok, I have added a judgment logic for multiple GPUs

**Review by wenmengzhou — APPROVED** [review]

File: tests/trainers/test_dreambooth_diffusion_trainer.py (hunk 7 above)
Comment (XDUWQ) [reply tests/trainers/test_dreambooth_diffusion_trainer.py:72]: Okay, pipeline has been added.

File: modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py (hunk 4 above)
Comment (XDUWQ) [reply modelscope/trainers/multi_modal/dreambooth_diffusion/dreambooth_diffusion_trainer.py:165]: This problem has been resolved

**Review by wenmengzhou — APPROVED** [review]

**Pull request merged**